File: 1250-longest-common-subsequence/1250-longest-common-subsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        m, n = len(text1), len(text2)
        if m > n:
            text1, text2 = text2, text1
            m, n = n, m

        pre = [0]*(m+1)
        cur = [0]*(m+1)
        for i in range(1, n+1):
            for j in range(1, m+1):
                if text1[j-1] == text2[i-1]:
                    cur[j] = pre[j-1]+1
                else:
                    cur[j] = max(cur[j-1], pre[j])

            pre = cur[:]
        
        return pre[-1]

File: 1250-longest-common-subsequence/1250-longest-common-subsequence.cpp
class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        int n = text1.length(), m = text2.length();
        vector<int> prev(m+1, 0);
        vector<int> curr(m+1, 0);

        for (int i=1; i<n+1; i++) {
            for (int j=1; j<m+1; j++) {
                if (text1[i-1] == text2[j-1]) {
                    curr[j] = prev[j-1]+1;
                } else {
                    curr[j] = max(prev[j], curr[j-1]);
                }
            }
            prev = curr;
        }

        return curr[m];
    }
};